plugins/postgres.php

```diff
@@ -0,0 +1,177 @@
+
+    public function __construct($type)
+    {
+        parent::__construct($type);
+
+        $extensions = get_loaded_extensions();
+        $extensions = array_flip($extensions);
+        if (!isset($extensions['pgsql'])) {
+            throw new RuntimeException('PGSQL extension is not enabled');
+        }
+
+    }
+
+    protected function url(): string
+    {
+        return "https://www.postgresql.org/";
+    }
+
+    protected function description(): string
+    {
+        return "The PostgreSQL object-relational database system provides reliability and data integrity.";
+    }
+
+    private function getConnection()
+    {
+        return $this->connection ?? pg_pconnect("host=127.0.0.1 port=$this->port user=$this->user");
+    }
+
+    // attempts to fetch info about engine and return it
+    protected function getInfo()
+    {
+
+        $connection = $this->getConnection();
+        $status = pg_connection_status($connection);
+        if ($status === PGSQL_CONNECTION_OK) {
+            $result = pg_query($connection, "SELECT name, setting FROM pg_settings;");
+            if ($result) {
+                $out = [];
+                while ($row = pg_fetch_row($result)) {
+                    $out[$row[0]] = $row[1];
+                }
+                $ret['variables'] = $out;
+            }
+
+            $result = pg_query($connection, "SHOW server_version;");
+            if ($result) {
+                $result = pg_fetch_all($result);
+                $ret['version'] = $result[0]['server_version'];
+            }
+        }
+
+
+        return $ret ?? false;
+    }
+
+    protected function appendType($info): string
+    {
+        return "";
+    }
+
+    protected function prepareQuery($query)
+    {
+        return $query; // no modifications required
+    }
+
+    // returns true when it's possible to connect to engine and it can run a simple query
+    protected function canConnect(): bool
+    {
+        $connection = $this->getConnection();
+        $status = pg_connection_status($connection);
+        if ($status === PGSQL_CONNECTION_OK) {
+            $result = pg_query($connection, "show all");
+            if ($result) {
+                return true;
+            }
+
+        } elseif (pg_ping($connection)) {
+            $result = pg_query($connection, "show all");
+            if ($result) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    protected function beforeQuery():void
+    {
+        $this->connection = $this->getConnection();
+
+        $timeout = self::$commandLineArguments['query_timeout'] * 1000;
+        pg_query($this->connection, "SET statement_timeout TO ".$timeout.";");
+    }
+
+    // runs one query against engine
+    // must respect self::$commandLineArguments['query_timeout']
+    // must return ['timeout' => true] in case of timeout
+    protected function testOnce($query)
+    {
+        if (!pg_connection_busy($this->connection)) {
```

Comment:
Try to avoid calling any functions other than only the needed one which runs the query. Take a look at https://github.com/db-benchmarks/db-benchmarks/blob/main/plugins/mysql.php#L69 as an example.

Reply:
done